nvme/opts: parse timeouts in humantime

tiagolobocastro opened the pull request:
ci(bors): add fake DCO for bors merge requests
    
    DCO is already validated on PR and bors itself does not need it as it's a bot..
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    fix(nvme/opts): remove outdated async parameter
    
    Remove unused NVME_QPAIR_CONNECT_ASYNC opts parameter and add
    generated uuid parameter.
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    fix(nvme/kato): adjust to spdk minimum
    
    SPDK minimum KATO is 10s, so setting 1s is pointless.
    Also the spec does suggest setting a sufficiently high KATO on a fabric, so perhaps
    we might want to consider increasing it in the future.
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    fix(iobuf): increase large iobuf pool size
    
    Addresses warning from spdk itself:
    mayastor::spdk:transport.c:282]
    The num_shared_buffers value (2048) is larger than the available iobuf pool size (1024).
     Please increase the iobuf pool sizes.

    Also reduce size to power of 2 - 1 as recommended in spdk ticket:
    It’s always best to specify a value for –n that is one less than a power of two (2^x -1)
    because of the way rings work in DPDK. You end up allocating a ring twice the size for
    the one extra element between 511 and 512 for example.
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    chore: silence flush not supported IO
    
    We don't keep any caching, so we don't actually need to make use of it anyway.
    TODO: do we need to even send flush?
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    chore: show starting configuration as info
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

---

    feat(nvme/opts): parse timeouts in humantime
    
    This allows us to specify, for example, NVME_TIMEOUT=60s
    
    Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
feat(nvme/opts): parse timeouts in humantime

This allows us to specify, for example, NVME_TIMEOUT=60s

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
chore: show starting configuration as info

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
chore: silence flush not supported IO

We don't keep any caching, so we don't actually need to make use of it anyway.
TODO: do we need to even send flush?

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

Abhinandan-Purkait commented on io-engine/src/subsys/config/mod.rs:
```diff
@@ -227,6 +227,6 @@ impl Config {
         assert!(self.iobuf_opts.set());
 
-        debug!("{:#?}", self);
+        info!("{:#?}", self);
```
Does this need to be info?

tiagolobocastro replied:
Yes, otherwise debugging SB we don't see the full config by default

Abhinandan-Purkait reviewed (approved):
LGTM

Abhinandan-Purkait commented on io-engine/src/subsys/config/opts.rs:
```diff
@@ -174,6 +175,74 @@ where
+                    value
+                }
+                Err(e) => {
+                    error!("Invalid value: {} (error {}) specified for {}. Reverting to default ({}{})", value, e, name, default, unit.units());
```
`Reverting to default value` ? Since we said the word value after Invalid

tiagolobocastro replied:
It follows existing message above (not shown on diff)
I'll change it for both

tiagolobocastro committed:
fix(iobuf): fixup large iobuf pool size

Addresses warning from spdk itself:
mayastor::spdk:transport.c:282]
The num_shared_buffers value (2048) is larger than the available iobuf pool size (1024).
 Please increase the iobuf pool sizes.

Also reduce size to power of 2 - 1 as recommended in spdk ticket:
It’s always best to specify a value for –n that is one less than a power of two (2^x -1)
because of the way rings work in DPDK. You end up allocating a ring twice the size for
the one extra element between 511 and 512 for example.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

dsharma-dc commented on io-engine/src/subsys/config/opts.rs:
```diff
@@ -668,7 +668,7 @@ impl Default for IoBufOpts {
         Self {
             small_pool_count: try_from_env("IOBUF_SMALL_POOL_COUNT", 8192),
-            large_pool_count: try_from_env("IOBUF_LARGE_POOL_COUNT", 1024),
+            large_pool_count: try_from_env("IOBUF_LARGE_POOL_COUNT", 2048),
```
This config might be performance sensitive. I'd recommend doing a sanity perf run with and without this change. Also, sometime back I was reading spdk issue 809, where it says 
> It’s always best to specify a value for –n that is one less than a power of two (2^x -1) because of the way rings work in DPDK. You end up allocating a ring twice the size for the one extra element between 511 and 512 for example.

So along with this, we can consider changing NVMF_TCP_NUM_SHARED_BUF from 2048 to 2047 as well.

tiagolobocastro replied:
Changed, and for my limited testing I see no perf impact.

dsharma-dc commented on io-engine/src/subsys/config/opts.rs:
```diff
@@ -668,7 +668,7 @@ impl Default for IoBufOpts {
         Self {
             small_pool_count: try_from_env("IOBUF_SMALL_POOL_COUNT", 8192),
-            large_pool_count: try_from_env("IOBUF_LARGE_POOL_COUNT", 1024),
+            large_pool_count: try_from_env("IOBUF_LARGE_POOL_COUNT", 2047),
```
Sorry if it wasn't clear from my earlier comment, but the -1 is to be done only for num_shared_buffers. For iobuf pool count it can remain as it is, the spdk defaults are full value too:
> #define IOBUF_DEFAULT_SMALL_POOL_SIZE	8192
#define IOBUF_DEFAULT_LARGE_POOL_SIZE	1024

The transport shared buffers is what's needed to be at -1

> #define SPDK_NVMF_TCP_DEFAULT_NUM_SHARED_BUFFERS 511

tiagolobocastro replied:
Right yes, done

tiagolobocastro committed:
fix(nvme/kato): adjust to spdk minimum

SPDK minimum KATO is 10s, so setting 1s is pointless.
Also the spec does suggest setting a sufficiently high KATO on a fabric, so perhaps
we might want to consider increasing it in the future.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro committed:
fix(io-engine/client): jsonrpc does not use subcommand

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented:
bors merge

tiagolobocastro commented:
bors cancel

tiagolobocastro committed:
fix(nvme/opts): remove outdated async parameter

Remove unused NVME_QPAIR_CONNECT_ASYNC opts parameter and add
generated uuid parameter.

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented:
bors merge

tiagolobocastro committed:
ci(bors): add fake DCO for bors merge requests

DCO is already validated on PR and bors itself does not need it as it's a bot..

Signed-off-by: Tiago Castro <tiagolobocastro@gmail.com>

tiagolobocastro commented:
bors merge

tiagolobocastro commented:
Had to enable DCO from any source...
bors merge
